tests/test_data.py

```diff
@@ -0,0 +1,187 @@
+from json import dumps
+import pytest
+from time import sleep
+
+from data import DataFactory, DataV2, DataV3
+import topics
+
+def test_data_factory():
+    assert isinstance(DataFactory.create("V2"), DataV2)
+    assert isinstance(DataFactory.create("V3"), DataV3)
+    with pytest.raises(NotImplementedError):
+        DataFactory.create(None)
+    with pytest.raises(NotImplementedError):
+        DataFactory.create("V9000")
+
+def to_query_string(python_dict):
+    """ Converts {key1: value1, key2: value2} to key1=value1&key2value2 """
+    return "&".join([f"{k}={v}" for k, v in python_dict.items()])
+
+def test_v2_sensor_parsing():
+    test_data = {
+        "gps": 1,
+        "gps_speed": 78.1,
+        "reed_velocity": 78.5,
+        "reed_distance": 118,
+        "power": 320,
+        "cadence": 100,
+    }
+
+    data = DataV2()
+    data.load_data(topics.DAS.data, to_query_string(test_data))
+
+    for key, value in test_data.items():
+        assert data[key] == value, f"Key {key} is not being set correctly"
+
+def test_v2_power_model_parsing():
+    recommended_sp_data = {
+        "rec_power": 500.0,
+        "rec_speed": 9.2772,
+        "zdist": 1219.0
+    }
+    max_speed_data = { "predicted_max_speed": 32.911 }
+    plan_name_data = { "plan_name": "default.pkl" }
+
+    data = DataV2()
+    data.load_data(topics.PowerModel.recommended_sp, to_query_string(recommended_sp_data))
+    data.load_data(topics.PowerModel.predicted_max_speed, to_query_string(max_speed_data))
+    data.load_data(topics.PowerModel.plan_name, to_query_string(plan_name_data))
+
+    for key, value in { **recommended_sp_data, **max_speed_data, **plan_name_data }.items():
+        assert data[key] == value, f"Key {key} is not being set correctly"
+
+def test_v3_messages():
+    message_packet = { "message": "testing 123" }
+    data = DataV2()
+    data.load_data(topics.DAShboard.receive_message, dumps(message_packet))
+    assert data.has_message()
+    assert data.get_message() == message_packet["message"]
+    sleep(data.message_duration)
+    assert not data.has_message()
```

Comment:
This test fails because I've written this test to comply with V3 MQTT spec, but the actual implementation was just ported over from what we have before, which is just sending the message as a plain string. monash-human-power/dashboard#67 also uses plain strings, although it's not merged so not too late.

Merging a failing test to master is probably bad, so should I update the implementation to use JSON, or the test to use a plain string?

Reply:
Having typed this out, it's made it clear that the implementation of this PR and monash-human-power/dashboard#67 should probably be updated to match the spec.